sw/host/opentitanlib/src/util/status.rs

```diff
@@ -0,0 +1,255 @@
+#[derive(Debug, serde::Serialize, serde::Deserialize, TryFromPrimitive, PartialEq, Eq)]
+#[repr(u32)]
+pub enum StatusCode {
+    Ok = bindgen::status::absl_status_code_kOk,
+    Cancelled = bindgen::status::absl_status_code_kCancelled,
+    Unknown = bindgen::status::absl_status_code_kUnknown,
+    InvalidArgument = bindgen::status::absl_status_code_kInvalidArgument,
+    DeadlineExceeded = bindgen::status::absl_status_code_kDeadlineExceeded,
+    NotFound = bindgen::status::absl_status_code_kNotFound,
+    AlreadyExists = bindgen::status::absl_status_code_kAlreadyExists,
+    PermissionDenied = bindgen::status::absl_status_code_kPermissionDenied,
+    ResourceExhausted = bindgen::status::absl_status_code_kResourceExhausted,
+    FailedPrecondition = bindgen::status::absl_status_code_kFailedPrecondition,
+    Aborted = bindgen::status::absl_status_code_kAborted,
+    OutOfRange = bindgen::status::absl_status_code_kOutOfRange,
+    Unimplemented = bindgen::status::absl_status_code_kUnimplemented,
+    Internal = bindgen::status::absl_status_code_kInternal,
+    Unavailable = bindgen::status::absl_status_code_kUnavailable,
+    DataLoss = bindgen::status::absl_status_code_kDataLoss,
+    Unauthenticated = bindgen::status::absl_status_code_kUnauthenticated,
+}
+
+// Safe abstraction over status_create
+fn status_create_safe(code: StatusCode, mod_id: u32, file: String, arg: i32) -> RawStatus {
+    // We do not expect an error since it is a valid String.
+    let file = CString::new(file).expect("CString::new failed");
+    unsafe {
+        // Safety: the function expects a valid readonly C-string which is exactly what
+        // CString:as_ptr() provides.
+        status_create(code as u32, mod_id, file.as_ptr(), arg)
+    }
+    // Note: file is dropped here so the C-string pointer is valid accross the function call.
+}
+
+// Convert an array of i8 to a string. This function will stop at first 0 (or at the
+// end of the array if it contains no zero).
+fn c_string_to_string(array: &[i8]) -> Result<String> {
+    // It might contain a lot of zeroes at the end which CString does not like and that we need to remove.
+    let s = array
+        .iter()
+        .map(|c| *c as u8)
+        .take_while(|c| *c != 0u8)
+        .collect::<Vec<_>>();
+    Ok(CString::new(s)?
+        .to_string_lossy() // returns a Cow
+        .to_string())
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+/// Hold a completely decoded status.
+pub struct Status {
+    // Module ID if present.
+    pub module_id: Option<String>,
+    // Status argument. Note that depending on whether this is an error status or ok
+    // status, the maximum value that can fit is different.
+    pub arg: i32,
+    // Error code of the status.
+    pub code: StatusCode,
+}
+
+// Safe abstraction over status_extract.
+impl TryFrom<RawStatus> for Status {
+    type Error = anyhow::Error;
+
+    fn try_from(status: RawStatus) -> Result<Status> {
```

Comment:
The idea was to avoid reimplementing all the bit manip and the filename to module ID derivation, as this would duplicate the information in two places.